### PR TITLE
Function layered integral

### LINKED ISSUE #19478: Create FunctionLayeredIntegral to integrate a function
## Reason
In Cardinal, it would be convenient to set an initial condition for fluid temperature based on simple bulk energy conservation, which requires a layered integral of a variable representing the power density. To get the dependency of ICs, UOs, and postprocessors right, this is easiest to achieve by adding a `FunctionLayeredIntegral` that simply integrates a function in layers (i.e. the same as a `LayeredIntegral`, but instead integrating a function instead of a variable).

## Design
Create a `FunctionElementIntegralUserObject` that can then be one of the base classes for a `FunctionLayeredIntegral`. 

## Impact
New feature.

### PR DESCRIPTION
**Pull request by aprilnovak**
Adds a `FunctionLayeredIntegral`, basically the same as `LayeredIntegral` except that it integrates a function instead of a variable. Needed for setting fluid temperature ICs in Cardinal based on bulk energy conservation.

Closes #19478

### CODE REVIEW

File: framework/include/userobject/FunctionElementIntegralUserObject.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ElementIntegralUserObject.h"
+
+class FunctionElementIntegralUserObject;
+
+template <>
+InputParameters validParams<FunctionElementIntegralUserObject>();
```
Comment (GiudGiud) [inline framework/include/userobject/FunctionElementIntegralUserObject.h:17]: let s no re-introduce this and use the new way for validparams

**Comment by GiudGiud** [comment]
@lindsayad should we make a FunctorLayeredIntegral ?
@aprilnovak how s your timeline on this? Functors include functions, material properties and variables, so the more general interface should do what you need too

**Comment by aprilnovak** [comment]
I need this feature soon. I don't know what functors are, so can't really comment on that. If someone could throw together/explain how to make a `FunctionLayeredIntegral` that does the same thing, that'd work for me.

**Comment by lindsayad** [comment]
> @lindsayad should we make a FunctorLayeredIntegral ?

Perhaps so. It would have made @aprilnovak's reimplementation here for functions unnecessary. However, since if we used the functor system with a function we would have to reinit an FE object with the qrule in order to get the quad points, when that is already done in the `ComputeUserObjectsThread` loop, it probably makes sense to keep what @aprilnovak has done here because it will be faster.

File: framework/include/userobject/FunctionElementIntegralUserObject.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ElementIntegralUserObject.h"
+
+class FunctionElementIntegralUserObject;
```
Comment (GiudGiud) [inline framework/include/userobject/FunctionElementIntegralUserObject.h:14]:
```suggestion
```
was for legacy params

File: framework/include/userobject/FunctionElementIntegralUserObject.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * This postprocessor computes a volume integral of a specified function.
```
Comment (GiudGiud) [inline framework/include/userobject/FunctionElementIntegralUserObject.h:17]:
```suggestion
 * This user object computes a volume integral of a specified function.
```

File: framework/src/userobject/FunctionElementIntegralUserObject.C [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FunctionElementIntegralUserObject.h"
+#include "Function.h"
+
+registerMooseObject("MooseApp", FunctionElementIntegralUserObject);
+
+defineLegacyParams(FunctionElementIntegralUserObject);
```
Comment (GiudGiud) [inline framework/src/userobject/FunctionElementIntegralUserObject.C:15]:
```suggestion
```

File: framework/src/userobject/FunctionLayeredIntegral.C [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FunctionLayeredIntegral.h"
+
+registerMooseObject("MooseApp", FunctionLayeredIntegral);
+
+defineLegacyParams(FunctionLayeredIntegral);
```
Comment (GiudGiud) [inline framework/src/userobject/FunctionLayeredIntegral.C:14]:
```suggestion
```

File: test/tests/userobjects/function_layered_integral/tests [hunk 6]
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues='#19478'
+  design='FunctionLayeredIntegral.md'
+
+  [function_layered_integral]
+    type = Exodiff
+    input = 'function_layered_integral.i'
+    exodiff = 'function_layered_integral_out.e'
+    requirement = 'The system shall integrate a function in layers'
```
Comment (GiudGiud) [inline test/tests/userobjects/function_layered_integral/tests:9]:
```suggestion
    requirement = 'The system shall be able to integrate a function over layers'
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**